parcels/nodes/PyNode.py

```diff
@@ -0,0 +1,601 @@
+        :arg id: optional, legacy parameter to set this object's ID
+        :arg data: data payload of this Node
+        :arg c_lib_register: a LibraryRegisterC object; available for compliance reasons with the NodeJIT sibling class
+        :arg idgen: an ID generator object used for ID registration- and de-registration for the Particle data payload
+        """
+        if prev is not None:
+            assert (isinstance(prev, Node))
+            self.set_prev(prev)
+        else:
+            self.reset_prev()
+        if next is not None:
+            assert (isinstance(next, Node))
+            self.set_next(next)
+        else:
+            self.reset_next()
+        if data is not None:
+            self.set_data(data)
+        else:
+            self.reset_data()
+        self.link()
+
+        assert idgen is not None, "Using Nodes requires to specify an ID generator (in order to release the ID on delete)."
+        self.idgen = idgen
+        self.registered = True
+
+    def __del__(self):
+        """
+        Node - Destructor
+        """
+        if self.data is not None:
+            try:
+                self.idgen.releaseID(self.data.id)
+            except:
+                pass
+        del self.data
+        self.unlink()
+        self.reset_data()
+
+    def __deepcopy__(self, memodict={}):
+        """
+        :returns a deepcopy of this very object
+        """
+        result = type(self)(prev=None, next=None, id=-1, data=None)
+        result.registered = True
+        result.next = self.next
+        result.prev = self.prev
+        result.data = self.data
+        return result
+
+    def link(self):
+        """
+        links this node to its neighbours, i.e. sets this object to be the successor of its predecessor and
+        sets this object to be the predecessor of its successor, so that we have a mutual connection in the form of:
+        prev <-> this <-> next
+        """
+        if self.prev is not None and self.prev.next != self:
+            self.prev.set_next(self)
+        if self.next is not None and self.next.prev != self:
+            self.next.set_prev(self)
+
+    def unlink(self):
+        """
+        removes this object from the linked chain of nodes, i.e. we set this predecessor's successor to be this successor
+        and set this successor's predecessor to this predecessor, so that the new connection is in the form of:
+        prev <-> next
+        """
+        if self.registered:
+            if self.prev is not None:
+                if self.next is not None:
+                    self.prev.set_next(self.next)
+                else:
+                    self.prev.reset_next()
+            if self.next is not None:
+                if self.prev is not None:
+                    self.next.set_prev(self.prev)
+                else:
+                    self.next.reset_prev()
+        self.reset_prev()
+        self.reset_next()
+        self.registered = False
+
+    def is_valid(self):
```

Comment:
Wouldn't it be more pythonic to call this `isvalid()`? (see `isnan`, `isinstance` etc)